### PR TITLE
Improve performance of JitCachedFunctionSearchStarted profiler hook.

### PR DESCRIPTION
**Pull request by macrogreg**
The loader hooks into the `JitCachedFunctionSearchStarted` profiler callback in order to selectively disable the NGen usage for one particular method that needs to be instrumented to inject the managed loader.

This change improves the performance of that callback:

* Superfluous allocations and string conversions are removed.

* Once the function id of the target method is known, a fast path is taken, that avoids metadata loading and string comparisons and uses just 2 integer comparisons instead. In an ad-hoc test, a 15 sec execution of the Profiler's Computer demo resulted in a total of 686 invocations of `HandleJitCachedFunctionSearchStarted`, out of which 677 (=98.7%) ended up using the fast path.

### CODE REVIEW

File: src/Shared-Src/Native/loader.cpp [hunk 1]
```diff
@@ -443,59 +445,90 @@ namespace shared
 
     HRESULT Loader::HandleJitCachedFunctionSearchStarted(FunctionID functionId, BOOL* pbUseCachedFunction)
     {
+        // Some of the logging in this method is a little too verbode even for debug.
+        // Turn it on here if you need it for investigations.
+        constexpr bool ExtraVerboseLogging = false;
```
Comment (gleocadie) [inline src/Shared-Src/Native/loader.cpp:450]: Generally we turn on Debug logging to investigate. Why having an extra one ?

Comment (gleocadie) [reply src/Shared-Src/Native/loader.cpp:450]: This one need to rebuild everything to activate it, you cannot activate it by env. var.... tha's not really useful in the long term

**Review by gleocadie — CHANGES_REQUESTED** [review]

File: src/Shared-Src/Native/loader.cpp (hunk 1 above)
Comment (macrogreg) [reply src/Shared-Src/Native/loader.cpp:450]: Agree. Previously it was in the code and commented out and @tonyredondo  recommended that I uncomment it for the investigation I made (which was to see how often this gets invoked).
I am happy to leave it there commented like it was. I only made the compile time switch because it is now in 2 places (fast path and normal path) and this allows turning this on at once.

What is your preference, @gleocadie ?

Comment (gleocadie) [reply src/Shared-Src/Native/loader.cpp:450]: Oh you mean the Debug statement that was commented and you replace with the `if (debugEnabled)`.  A bit hard to follow :)

To be honest, do it your way, but I think that we should have a proper tooling/mechanism to investigate easily. Tomorrow, when running in the customer environment, I thing asking to re-run with specific flags to get detailed information is a better approach that sending a patched version.
Another ticket in our backlog :)

**Review by gleocadie — APPROVED** [review]

**Comment by tonyredondo** [comment]
Before merging this, we need to ensure one thing…

We must be sure the same type.method doesn’t get searched again using a different functionId due a different AppDomainId.

File: src/Shared-Src/Native/loader.cpp (hunk 1 above)
Comment (macrogreg) [reply src/Shared-Src/Native/loader.cpp:450]: @gleocadie There two separate points here. I think you are right on both, I just wanna keep them separate. :)

a) The `Debug(..)` statement was commented out before the change (think, if was compile-time-switched). How that there are two _related_ Debug statements in this method I replaced the compile-time-switch based on comments with another compile-time-switch based on a method-scoped `constexpr`.

b) The `Debug(..)` statements in the entire class were taking cycles even if debug logging was disabled, because or the parameter formatting issue we discussed earlier. I addressed it with `logDebugIsEnabled` flag (earlier PR).

Point (b) definitely needs some improvement./refactoring in the mid-term, as discussed.

Point (a) is minor and local, I do not mind what way we go there. I think it's not worth putting much effort into. Here, I just explained why I did what I did, but I'd be happy to do it in any other way. :)

**Comment by macrogreg** [comment]
> Before merging this, we need to ensure one thing…
> 
> We must be sure the same type.method doesn’t get searched again using a different functionId due a different AppDomainId.

Very true. Thanks, @tonyredondo !  I started doing that and I have some questions (see Slack).

**Comment by gleocadie** [comment]
It would be great to have those questions somewhere public where everyone can learn about it, no?  Maybe in this discussion, that would benefit everyone

**Comment by macrogreg** [comment]
> Before merging this, we need to ensure one thing…
> 
> We must be sure the same type.method doesn’t get searched again using a different functionId due a different AppDomainId.

Excellent point. Thank you for bringing it us and for the help looking at the logs.
We concluded that the method being instrumented here (`System.AppDomain.IsCompatibilitySwitchSet`) is in `mscorlib`, which is AD-neutral, so we will only encounter it once in `HandleJitCachedFunctionSearchStarted`.

I'll merge the PR and reach out to you for a chat about some background questions I have on the loader that came up for me in the process of looking at this.

Cheers!

(FYI @tonyredondo ,  @gleocadie )
